### DISCUSSION TITLE
`PI score calculation failed`

### DISCUSSION THREAD

**Discussion by jsquaredosquared** [KosinskiLab/AlphaPulldown#402] [Q&A] [answered]
In the `statistics.csv` file, the following fields are empty for all predictions:

Num_intf_residues | Polar | Hydrophobic | Charged | contact_pairs | sc | hb | sb | int_solv_en | int_area | pi_score
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --

Why are these fields empty?

Looking at the log file for the `compute_stats` rule, I get the following error:

```
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Provided resources: mem_mb=8000, mem_mib=7630, disk_mb=1000, disk_mib=954, walltime=1440, attempt=1
Conda environments: ignored
Select jobs to execute...

[Mon Aug 26 01:16:18 2024]
rule compute_stats:
    input: /scr/user/jeffrey.o/alphapulldown_output/predictions/Q6BDS2_P51149/completed_fold.txt
    output: /scr/user/jeffrey.o/alphapulldown_output/predictions/Q6BDS2_P51149/statistics.csv
    jobid: 0
    reason: Forced execution
    wildcards: fold=Q6BDS2_P51149
    resources: mem_mb=8000, mem_mib=7630, disk_mb=1000, disk_mib=954, tmpdir=/tmp, walltime=1440, attempt=1, partition=cpu-epyc-genoa,cpu-epyc

Activating singularity image /home/user/jeffrey.o/AlphaPulldownSnakemake/.snakemake/singularity/3f20617ccba864758b2a437ef2fde35c.simg
INFO:    Could not find any nv files on this host!
INFO:    Converting SIF file to temporary sandbox...
I0826 01:18:23.496333 22361792987968 get_good_inter_pae.py:120] now processing Q6BDS2_P51149
E0826 01:18:31.441169 22361792987968 pdb_analyser.py:173] PI score calculation failed: Command '['/bin/bash', '-c', 'source activate pi_score && export PYTHONPATH=/software:$PYTHONPATH && python /software/pi_score/run_piscore_wc.py -p /tmp/tmp.ubjPZQDmEl/Q6BDS2_P51149/ranked_0.pdb -o /tmp/tmp.ubjPZQDmEl/pi_score_outputs/Q6BDS2_P51149 -s 2 -ps 10']' returned non-zero exit status 1.. Command: ['/bin/bash', '-c', 'source activate pi_score && export PYTHONPATH=/software:$PYTHONPATH && python /software/pi_score/run_piscore_wc.py -p /tmp/tmp.ubjPZQDmEl/Q6BDS2_P51149/ranked_0.pdb -o /tmp/tmp.ubjPZQDmEl/pi_score_outputs/Q6BDS2_P51149 -s 2 -ps 10']. Error: Last system error message: No such file or directory
 SC:   Open failed: File: /tmp/root/sc_tmp.19787                                                                                     

W0826 01:18:31.442494 22361792987968 pdb_analyser.py:192] Skipping interface due to unexpected format: None
E0826 01:18:31.442568 22361792987968 pdb_analyser.py:319] An unexpected error occurred while merging the PI score data: local variable 'pi_score_files' referenced before assignment
W0826 01:18:31.442638 22361792987968 pdb_analyser.py:192] Skipping interface due to unexpected format: None
I0826 01:18:34.871306 22361792987968 get_good_inter_pae.py:168] done for Q6BDS2_P51149 1 out of 1 finished.
INFO:    Cleaning up image...
[Mon Aug 26 01:18:56 2024]
Finished job 0.
1 of 1 steps (100%) done
```

What can be done to fix this?

**Comment by DimaMolod** [comment]
Hi @jsquaredosquared 
it could be that pi-score fails because no interface has been found.
Could you enter your container:
`singularity shell /home/user/jeffrey.o/AlphaPulldownSnakemake/.snakemake/singularity/3f20617ccba864758b2a437ef2fde35c.simg`
and try running this command directly:
`source activate pi_score && \
export PYTHONPATH=/software:$PYTHONPATH && \
python /software/pi_score/run_piscore_wc.py \
-p /tmp/tmp.ubjPZQDmEl/Q6BDS2_P51149/ranked_0.pdb \
-o /tmp/tmp.ubjPZQDmEl/pi_score_outputs/Q6BDS2_P51149 \
-s 2 \
-ps 10
`
then you should get a clearer error message. 
Also, please open your Q6BDS2_P51149 model, e.g., in PyMol, and check whether there is an interface.
Please let me know about your findings.
Cheers
d

**Reply by jsquaredosquared** [reply]
I tried running the command directory but got the following error:

```
Checking local paths
*****Selecting model for training*****
*****Setting output directory for results*****
Traceback (most recent call last):
  File "/software/pi_score/run_piscore_wc.py", line 242, in <module>
    main()
  File "/software/pi_score/run_piscore_wc.py", line 65, in main
    os.mkdir(args.out)
OSError: [Errno 2] No such file or directory: '/tmp/tmp.ubjPZQDmEl/pi_score_outputs/Q6BDS2_P51149'
```

I also tried viewing the Q6BDS2_P51149 model, but I'm not sure how to tell whether there is an interface or not:

jobs | iptm_ptm | iptm | pDockQ/mpDockQ | average_interface_pae | average_interface_plddt | binding_energy | interface | Num_intf_residues | Polar | Hydrophobic | Charged | contact_pairs | sc | hb | sb | int_solv_en | int_area | pi_score
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Q6BDS2_P51149 | 0.793503560689028 | 0.818551589717804 | 0.230605479214527 | 8.12593984962407 | 54.7868715083799 | -66814.4808615352 | B_C |   |   |   |   |   |   |   |   |   |   |  

![image](https://github.com/user-attachments/assets/66959fc6-c86b-471d-bad4-4480f2b9931d)

**Reply by jsquaredosquared** [reply]
@DimaMolod do you have any other suggestions? 🥲

I have run the program several times. As far as I can tell, there is apparently an interface (B_C) yet the PI score calculation and other outputs fail.

**Reply by DimaMolod** [reply]
Hi @jsquaredosquared, and thank you for pointing this out! 
This indeed looks like a bug that we will fix as soon as possible.
You could try using the older version of the analysis pipeline image:
https://hub.docker.com/r/kosinskilab/fold_analysis/tags
or maybe just wait until we fix that :-)

**Reply by jsquaredosquared** [reply]
All right. I will patiently await your fix :D

**Reply by DimaMolod** [reply]
Hi @jsquaredosquared and I am sorry for the delay. 
Please feel free to update the docker images and run again, the problem is fixed now. 
I also have less positive news: we are going to remove ccp4 package from the analysis pipeline image very soon due to some legal issues. But we will also update the documentation and explain how AP users could build the image themselves and keep using it with our snakemake pipeline.

**Reply by jsquaredosquared** [reply]
> I am sorry for the delay. 

No worries. Thanks for your hard work :)

> we are going to remove ccp4 package

Sucks that you have to remove ccp4. I hope it won't be too difficult to incorporate it.

How helpful is it? Is it an absolutely necessary component, or can AlphaPulldown work well without it?

**Reply by DimaMolod** [reply, marked as answer]
No, it's not necessary at all. Without ccp4 it will all work up until the issue you reported here - these columns in the table will be always empty:

Num_intf_residues | Polar | Hydrophobic | Charged | contact_pairs | sc | hb | sb | int_solv_en | int_area | pi_score
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --

I hope it will be easy enough to install ccp4 within the container and run AP as before.
Thanks again for your report and happy modeling :-)

**Reply by jsquaredosquared** [reply]
Hi @DimaMolod , 

Sorry to bother you again. This time, for most of the predictions, the columns were filled in except for the pi_score, which is completely missing in the output file :/

![image](https://github.com/user-attachments/assets/144824de-4df5-42dd-ad10-ca53a5c0228c)

**Reply by DimaMolod** [reply]
Hi @jsquaredosquared, and sorry for this recurring issue. I will look again into this. 
Meanwhile, could you try running the singularity image manually for your predictions?
[Here](https://github.com/KosinskiLab/AlphaPulldown?tab=readme-ov-file#03-installation-for-the-downstream-analysis-tools), you can find how to download the image (please use the jax0.4 version), and [here](https://github.com/KosinskiLab/AlphaPulldown?tab=readme-ov-file#create-results-table) are the instructions on how to run it.
Please let me know if this helps

**Reply by jsquaredosquared** [reply]
Thanks @DimaMolod ,

After running it manually I was able to get the PI score :D

![image](https://github.com/user-attachments/assets/96c27350-fade-419c-8d65-6a38e3f22079)
